### ISSUE TITLE
OOM on loadParameters + saveParameters

### ISSUE THREAD

**Issue by pkese** [DiffSharp/DiffSharp#269]
While training, I'm saving model parameters every once in a while and I've found out that  
if I save a model that was previously loaded from file using `model.loadParameters`,  
then, while trying to write it with `model.saveParameters` it produces gigabytes of output file  
and the process gets killed due to memory exhaustion.

This does not happen if the model parameters weren't prevously loaded from file  
(i.e. if I start training with an uninitialized model).

I've tried this using DotNet 5 with both the current Nuget build as well as current GutHub 'dev' branch.  
I've spent some time figuring out what's going on, but ran out of ideas.

It doesn't matter if I keep replacing the same file or generate a new file name each time I write it - the result is the same.

**Comment by pkese** [comment]
It seems this doesn't happen always.  
I've tried to write some trivial code which I could attach to this ticket, but it seems it won't reproduce on trivial example.

**Comment by pkese** [comment]
I've tried to replace `model.saveParameters` with `model.clone()` + `model.noDiff()`+ `model.saveParameters`

The result was that it managed to save parameters without OOM at least on the first invocation,  
but on second loop, I got an exception trace:
```
Unhandled exception. System.Exception: Cannot load from file. "Binary stream '0' does not contain a valid BinaryHeader. Possible causes are invalid stream or object version change between serialization and deserialization."
   at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1433.Invoke(String message) in F:\workspace\_work\1\s\src\fsharp\FSharp.Core\printf.fs:line 1433
   at DiffSharp.Util.ExtensionAutoOpens.loadBinary[T](String fileName) in /home/ubuntu/src/DiffSharp/src/DiffSharp.Core/Extensions.fs:line 179
   at DiffSharp.Model.Model.load(String fileName) in /home/ubuntu/src/DiffSharp/src/DiffSharp.Core/Model.fs:line 274
   at DiffSharp.Model.Model.clone() in /home/ubuntu/src/DiffSharp/src/DiffSharp.Core/Model.fs:line 280
   at Fit.fit[a](a data) in /home/ubuntu/src/dom-control/Fit.fs:line 151
   at Program.main(String[] argv) in /home/ubuntu/src/dom-control/Program.fs:line 63
```
So it fails when `clone` calls `ExtensionAutoOpens.loadBinary` to deserialize binary stream that was used for cloning.  
Apparently that serialized binary stream was huge - memory usage at that point was about 4GB but it didn't OOM.

**Comment by pkese** [comment]
It turned out that I could avoid the problem simply by calling
`model.parameters.noDiff().save`  
instead of
`model.saveParameters`
and thus avoid gradient stuff getting saved along with parameters.

It works for my case, but I don't know if this is a correct solution for general use.

**Comment by dsyme** [comment]
Cool thanks for reporting this problem

**Comment by gbaydin** [comment]
@pkese thank you for sharing this. The description looks like the problem you encountered happened when the parameters were saved during a differentiation operation. 

I experimented with similar scenarios in my system and made changes in the model code that ensures derivatives are note saved by default when calling `Model.save` or `Model.saveState`. If someone needs to save and load the full derivative state, they can do so using an optional argument. 

This is implemented in #381 with corresponding unit tests.
